src/Swift.Runtime/src/Swift/Data.cs

```diff
@@ -1,10 +1,119 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Swift;
+using Swift.Runtime;
+
+
 /// <summary>
-/// Represents Foundation.Data type.
+/// Represents Swift Foundation.DataProtocol in C#.
+/// </summary>
+public interface ISwiftDataProtocol { }
+
+/// <summary>
+/// Represents Swift Foundation.ContiguousBytes in C#.
 /// </summary>
+public interface ISwiftContiguousBytes { }
 
-// TODO: https://github.com/dotnet/runtimelab/issues/2992
-public struct Data
+/// <summary>
+/// Represents Foundation.Data type.
+/// </summary>
+public unsafe struct Data : ISwiftObject
 {
     private long _flags;
     private IntPtr _object;
+
+    private static nuint _payloadSize = SwiftObjectHelper<Data>.GetTypeMetadata().Size;
+
+    private static Dictionary<Type, string> _protocolConformanceSymbols;
+
+    static Data()
+    {
+        _protocolConformanceSymbols = new Dictionary<Type, string> {
+            { typeof(ISwiftDataProtocol), "$s10Foundation4DataVAA0B8ProtocolAAMc" },
+            { typeof(ISwiftContiguousBytes), "$s10Foundation4DataVAA15ContiguousBytesAAMc" },
+        };
+    }
+
+    static TypeMetadata ISwiftObject.GetTypeMetadata()
+    {
+        return TypeMetadata.Cache.GetOrAdd(typeof(Data), _ => PInvoke_getMetadata());
+    }
+
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+    [DllImport(KnownLibraries.SwiftFoundation, EntryPoint = "$s10Foundation4DataVMa")]
+    public static unsafe extern TypeMetadata PInvoke_getMetadata();
+
+    static ISwiftObject ISwiftObject.NewFromPayload(SwiftHandle handle)
+    {
+        return *(Data*)handle;
+    }
+
+    IntPtr ISwiftObject.MarshalToSwift(IntPtr swiftDest)
+    {
+        var metadata = SwiftObjectHelper<Data>.GetTypeMetadata();
+        unsafe
+        {
```

Comment:
Why do we need unsafe here if the entire class is unsafe?

Reply:
Removed `unsafe` from the type.